gstreamer-pbutils/src/discoverer_stream_info.rs

```diff
@@ -0,0 +1,36 @@
+// Copyright (C) 2018 Thiago Santos <thiagossantos@gmail.com>
+//                    Sebastian Dröge <sebastian@centricular.com>
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use DiscovererStreamInfo;
+use DiscovererStreamInfoExt;
+
+pub struct DiscovererStreamInfoIterator {
+    stream_info: Option<DiscovererStreamInfo>
+}
+
+impl Iterator for DiscovererStreamInfoIterator {
+    type Item = DiscovererStreamInfo;
+
+    fn next(&mut self) -> Option<DiscovererStreamInfo> {
+        let current = self.stream_info.take();
+        self.stream_info = match &current {
+            &Some(ref c) => c.get_next(),
+            &None => None
+        };
+        current
+    }
+}
+
+impl DiscovererStreamInfo {
+    pub fn next_iter(&self) -> DiscovererStreamInfoIterator {
```

Comment:
And one for prev maybe. You can just store a boolean for the direction in the iterator, and it will then be optimized away by the compiler.

Also you could call this just Iterator or Iter, and not re-export it from src/lib.rs

Reply:
Not sure I understood this `Also you could call this just Iterator or Iter, and not re-export it from src/lib.rs`

I'll re-read on the prelude and exporting to see how this works and understand why we can do that.

Reply:
You would then have `gst_pbutils::DiscovererStreamInfo` and `gst_pbutils::discoverer_stream_info::Iter`. So that would not conflict with anything then, and as users generally don't have to name the type but only get it from the function, there would not be any inconvenience.

Check the caps iterator in GStreamer for example.